google/cloud/forseti/services/inventory/storage.py

```diff
@@ -571,25 +606,24 @@ def update(self, resource):
         if self.readonly:
             raise Exception('Opened storage readonly')
 
-        Inventory.update_resource(resource)
-        self.buffer.add(resource._row)
-
-    def read(self, key):
-        """Read a resource from the storage.
-
-        Args:
-            key (str): Key of the object to read.
-
-        Returns:
-            object: Row object read from database.
-        """
-
         self.buffer.flush()
-        return (
-            self.session.query(Inventory)
-            .filter(Inventory.index == self.index.id)
-            .filter(Inventory.key == key)
-            .one())
+
+        try:
+            new_rows = Inventory.from_resource(self.index, resource)
+            old_rows = self._get_resource_rows(resource.key())
+
+            new_dict = {row.type_class : row for row in new_rows}
+            old_dict = {row.type_class : row for row in old_rows}
+
+            for type_class in InventoryTypeClass.SUPPORTED_TYPECLASS:
+                if type_class in new_dict:
```

Comment:
Is there ever a case where there are rows in old_dict that are not in new_dict that should be deleted?

Reply:
We don't have any case that there are rows in old_dict should be deleted. This is why we want delete information here.
There are 3 rows returns at most:
resource, iam_policy, gcs_policy
Right now the update is only used when adding error logs. 
Only resource has been updated when adding error logs.